sdks/java/extensions/zetasketch/src/main/java/org/apache/beam/sdk/extensions/zetasketch/HllCountMergePartialFn.java

```diff
@@ -0,0 +1,151 @@
+   */
+  static final class HyperLogLogPlusPlusWrapper<T> {
+    @Nullable private HyperLogLogPlusPlus<T> hll;
+
+    private HyperLogLogPlusPlusWrapper(@Nullable HyperLogLogPlusPlus<T> hll) {
+      this.hll = hll;
+    }
+  }
+
+  /**
+   * Coder for the {@link HyperLogLogPlusPlusWrapper} class with generic type parameter {@code T}.
+   */
+  private static final class HyperLogLogPlusPlusWrapperCoder<T>
+      extends AtomicCoder<HyperLogLogPlusPlusWrapper<T>> {
+
+    private static final HyperLogLogPlusPlusWrapperCoder<?> INSTANCE =
+        new HyperLogLogPlusPlusWrapperCoder<>();
+
+    private final Coder<HyperLogLogPlusPlus<T>> nullableHllCoder =
+        NullableCoder.of(HyperLogLogPlusPlusCoder.of());
+
+    // Generic singleton factory pattern; the coder works for all HyperLogLogPlusPlusWrapper objects
+    // at runtime regardless of type T
+    @SuppressWarnings("unchecked")
+    static <T> HyperLogLogPlusPlusWrapperCoder<T> of() {
+      return (HyperLogLogPlusPlusWrapperCoder<T>) INSTANCE;
+    }
+
+    @Override
+    public void encode(HyperLogLogPlusPlusWrapper<T> value, OutputStream outStream)
+        throws IOException {
+      nullableHllCoder.encode(value.hll, outStream);
+    }
+
+    @Override
+    public HyperLogLogPlusPlusWrapper<T> decode(InputStream inStream) throws IOException {
+      return new HyperLogLogPlusPlusWrapper<T>(nullableHllCoder.decode(inStream));
+    }
+
+    @Override
+    public void verifyDeterministic() throws NonDeterministicException {
+      nullableHllCoder.verifyDeterministic();
+    }
+  }
+
+  @Override
+  public Coder<HyperLogLogPlusPlusWrapper<T>> getAccumulatorCoder(
+      CoderRegistry registry, Coder<byte[]> inputCoder) {
+    return HyperLogLogPlusPlusWrapperCoder.of();
+  }
+
+  @Override
+  public HyperLogLogPlusPlusWrapper<T> createAccumulator() {
+    return new HyperLogLogPlusPlusWrapper<>(null);
+  }
+
+  @Override
+  public HyperLogLogPlusPlusWrapper<T> addInput(
+      HyperLogLogPlusPlusWrapper<T> accumulator, byte[] input) {
+    if (accumulator.hll == null) {
+      @SuppressWarnings("unchecked")
+      HyperLogLogPlusPlus<T> hll = (HyperLogLogPlusPlus<T>) HyperLogLogPlusPlus.forProto(input);
+      accumulator.hll = hll;
+    } else {
+      accumulator.hll.merge(input);
+    }
+    return accumulator;
+  }
+
+  @Override
+  public HyperLogLogPlusPlusWrapper<T> mergeAccumulators(
+      Iterable<HyperLogLogPlusPlusWrapper<T>> accumulators) {
+    HyperLogLogPlusPlusWrapper<T> merged = createAccumulator();
+    for (HyperLogLogPlusPlusWrapper<T> accumulator : accumulators) {
+      if (accumulator.hll == null) {
+        continue;
+      }
+      if (merged.hll == null) {
+        // Cannot set merged.hll to accumulator.hll directly because we shouldn't mutate accumulator
+        @SuppressWarnings("unchecked")
+        HyperLogLogPlusPlus<T> hll =
+            (HyperLogLogPlusPlus<T>)
+                HyperLogLogPlusPlus.forProto(accumulator.hll.serializeToProto());
```

Comment:
Why doing a `serializeToProto` then immediately `forProto` (which I think is a `fromProto`). Can we directly do merged.hll = accumulator.hll?

Reply:
Reason is mentioned in the comment above:
`// Cannot set merged.hll to accumulator.hll directly because we shouldn't mutate accumulator`

If you are asking for reasons why we shouldn't mutate `accumulator`, it is related to a design flaw in our `CombineFn` function. The result of that is, in short, people who are implementing `mergeAccumulators` need to comply with this instruction:
https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Combine.java#L406

Reply:
Ah I see. Yeah didn't connect the comment with these links of code but now it makes much more sense.

Reply:
Would it make sense to move the comment to right on top of line 143? I was asking myself the exactly same thing as @amaliujia, and also overlooked the comment further up :) 
P.S. the serialize-then-deserialize step is fairly expensive, if we ever find a workaround for it in the future.